libecl/include/ert/ecl/ecl_type.h

```diff
@@ -0,0 +1,123 @@
+/*
+   Copyright (C) 2017  Statoil ASA, Norway.
+
+   The file 'ecl_type.h' is part of ERT - Ensemble based Reservoir Tool.
+
+   ERT is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   ERT is distributed in the hope that it will be useful, but WITHOUT ANY
+   WARRANTY; without even the implied warranty of MERCHANTABILITY or
+   FITNESS FOR A PARTICULAR PURPOSE.
+
+   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
+   for more details.
+*/
+
+#ifndef ERT_ECL_TYPE_H
+#define ERT_ECL_TYPE_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define ECL_STRING10_LENGTH 10    // Initial guess that the type C010 type is a10 character string.
+#define ECL_STRING8_LENGTH   8
+#define ECL_TYPE_LENGTH      4
+
+/*****************************************************************/
+/*
+   Observe that these type identidiers are (ab)used in both the rms and
+   ert/enkf libraries in situations where ECLIPSE is not at all involved.
+
+   Regarding the type ECL_C010_TYPE: In an SMSPEC file generated by IX
+   there was a keyword with header:
+
+      NAMES   316   C010
+
+   i.e. the type identifier was the string 'C010'. Have not found any
+   further documentation of that type, but assuming it is a 10
+   character long string the file loads ok. The ECL_C010_TYPE is
+   implemented on a very loose basis it is not clear if the
+   ten-character-long-string implementation makes sense, or if that
+   just happened to work for the one example we have encountered so
+   far. For this reason the support for this keyword type is very
+   limited:
+
+     - keywords of this type are not propertly internalized; they are
+       just cleanly skipped by ecl_kw/ecl_file implementation when
+       loading a file.
+
+     - it is not possible to instantiate a keyword of this type with
+       ecl_kw_alloc( ).
+
+     - The type is not exported to Python.
+*/
+
+typedef enum {
+  ECL_CHAR_TYPE   = 0,
+  ECL_FLOAT_TYPE  = 1,
+  ECL_DOUBLE_TYPE = 2,
+  ECL_INT_TYPE    = 3,
+  ECL_BOOL_TYPE   = 4,
+  ECL_MESS_TYPE   = 5,
+  ECL_C010_TYPE   = 6   // See comment immediately above about the limited support of this type.
+} ecl_type_enum;
+
+#define ECL_TYPE_ENUM_DEFS {.value = 0 , .name = "ECL_CHAR_TYPE"}, \
+{.value = 1 , .name = "ECL_FLOAT_TYPE"} ,                          \
+{.value = 2 , .name = "ECL_DOUBLE_TYPE"},                          \
+{.value = 3 , .name = "ECL_INT_TYPE"},                             \
+{.value = 4 , .name = "ECL_BOOL_TYPE"},                            \
+{.value = 5 , .name = "ECL_MESS_TYPE"}
+
+#define ECL_TYPE_ENUM_SIZE 6
+
+
+struct ecl_type_struct {
+    const ecl_type_enum type;
+
+    // In case of ECL_STRING_TYPE:
+    // length denotes the number of characters in the string
+    const size_t length; 
```

Comment:
If we introudcue this structure I think we should use it consistently - i.e. the `sizeof_ctype` field should be removed from the `struct ecl_kw_struct` - and:

1. `length` should be renamed `element_size` or something like that.
2. `element_size` should be correctly initialized for all the types, i.e.

```
#define ECL_INT (ecl_data_type) {.type = ECL_INT_TYPE, .element_size = 4}
```